### ISSUE TITLE
Debugger doesn't work for web tests

### ISSUE THREAD

**Issue by alanknight-wk** [Dart-Code/Dart-Code#3659]
**Describe the bug**
If you manage to get a configuration set up to run tests for the web (e.g. [like this](https://github.com/Dart-Code/Dart-Code/issues/3311)) you don't get a debugger. Neither breakpoints nor explicit debugger() statements get it to pause in the VS Code tools. If you add a --pause-after-load to the tool arguments then you can make it pause at the beginning, set breakpoints in Chrome Devtools, and run, and those will also respect debugger() statements, but it still doesn't show up in VS Code.

The root cause of this seems likely to be that package:test runs tests in a separate isolate (per test? one in total?) and VS Code is not connecting to it. That's also true on the VM, but there might be differences between the two or a flaw in the debug adaptor or DWDS that is causing it not to connect.

**To Reproduce**
In https://github.com/alanknight-wk/web_tests_with_vscode , click the debug button anywhere in foo_test.dart.
Attached is a log of running with a breakpoint set in VS Code, which completes without ever showing the debugger. But you can also uncomment the debugger() statement, in which case it never completes in VS Code. Or you can add a --pause-after-load to the launch config and pause it in Chrome DevTools, but nothing will ever show up in Vs Code until you let it complete.

**Versions (please complete the following information):**
 - VS Code version: 1.60.2
 - Dart extension version: 3.27.2
 - Dart/Flutter SDK version: 2.13.4

**Comment by alanknight-wk** [comment]
[dart-code-log.txt](https://github.com/Dart-Code/Dart-Code/files/7488856/dart-code-log.txt)

**Comment by DanTup** [comment]
@jakemac53 @natebosch does `package:test` support debugging tests from the editor through dwds? Right now I think in VS Code we're debugging VM tests because we just attach to the VM running the `package:test` code, and it spawns the tests in another isolate (which we can debug).

However when running browser tests, I presume we're still connecting to that VM running `package:test`, which is then spawning a browser to run the tests. I think we'd need to connect the IDE to a dwds instance that's connected to Chrome, but I don't know if that's currently possible (I can't find a way to get a VM Service URI that might be this).

**Comment by DanTup** [comment]
@jakemac53 @natebosch do you know the answer to the above? Is it possible to debug tests through dwds (eg. from the editor) when running in a browser?

**Comment by jakemac53** [comment]
No, package:test does not support this. You should be able to go through webdev to do it though.

**Comment by DanTup** [comment]
@jakemac53 I don't suppose that can be done in a way that's compatible with `package:test` (eg. running tests by name/location, and getting the json reporter output)?

**Comment by DanTup** [comment]
Also - it seems like there's no `webdev test` (only build/server/daemon). Does this mean `build_runner` should be used? (I'm not entirely familiar with the relationship between webdev and build_runner).

**Comment by jakemac53** [comment]
Weird that webdev doesn't have a test command... lol.  You could run tests by just doing `webdev serve test` and then loading and running them directly in the browser. But that won't support all the command line options and probably won't really support IDE integration well either.

One thing you could try is running `webdev serve test` in one process, and then `dart test --pub-serve=<port>` (only pass the port that webdev is running on). That uses some old support for loading tests over pub, which webdev should be compatible with?

But this really isn't a well supported option (and we have considered dropping support for the `--pub-serve` argument a few times...

**Comment by DanTup** [comment]
> we have considered dropping support for the `--pub-serve` argument a few times...

And why _didn't_ you go ahead with it? Right now I don't have any support for web tests in VS Code. It'd be nice to add something, but if the future of that is uncertain it might not be best to use it. If someone else is depending on it that makes it likely to hang around, maybe it is.

Would adding a `webdev test` command make sense? Should VS Code be using `webdev` (versus `build_runner` or something else)? I've never fully understood the relationship between these, so VS code currently has some webdev bits (running/debugging web apps - but not tests), and some build_runner bits (commands to build/run/serve).

**Comment by jakemac53** [comment]
> And why _didn't_ you go ahead with it?

I think just because it hasn't been causing us a lot of maintenance overhead, so nobody has bothered to, and it is potentially useful for cases like this. But I don't think it's even tested any more and if it was causing us headaches we would likely delete it.

If however it ends up working for debugging tests, that might justify us keeping it around and maintaining it.

> Would adding a `webdev test` command make sense?

It seems like it would to me, yes. But I don't know the amount of work or priority that would be placed on it. There is a `build_runner test` command but it doesn't support the debug service protocol (webdev adds that support). cc @kevmoo

**Comment by jakemac53** [comment]
> I've never fully understood the relationship between these, so VS code currently has some webdev bits (running/debugging web apps - but not tests), and some build_runner bits (commands to build/run/serve).

Webdev is just supposed to be a wrapper around build_runner (or any other future tool) which is stable. It was really built primarily for IDE plugins as a stable interface to talk to.

**Comment by evanweible-wf** [comment]
If `webdev` is the answer here, does that mean it would only be able to run/debug tests in the browser? In other words, would VS Code need to run separate processes for debugging tests on web vs VM? In particular I'm thinking of projects that maintain platform-agnostic code and want to test on multiple platforms, or projects that have VM code and web code to test.

I bring this up because there is already an unfortunate DX segmentation between plain Dart projects and Dart projects that use build_runner. In a [comment thread a while back](https://github.com/dart-lang/build/pull/2415#issuecomment-530114943), it sounded like there was an eventual goal to get back to the point where `dart test` would be capable of running tests regardless of whether build_runner was being used (and ideally regardless of the target platform).

In general, I'm just hoping that we could get to a point where there's less differentiation in the DX for web-based projects than there is now.

**Comment by jakemac53** [comment]
> If `webdev` is the answer here, does that mean it would only be able to run/debug tests in the browser?

You could actually use webdev to run the VM tests as well. It can serve up normal dart files also (it needs to for source maps to work anyways). It just delegates to `build_runner` which is platform agnostic.

I don't know that it's really a great solution though - it's going to be noticeably slower than loading from disk. And it requires build_runner to be in the project so it wouldn't work for all projects.

> In a [comment thread a while back](https://github.com/dart-lang/build/pull/2415#issuecomment-530114943), it sounded like there was an eventual goal to get back to the point where `dart test` would be capable of running tests regardless of whether build_runner was being used (and ideally regardless of the target platform).

We could do this still, although I think nowadays we might just sidestep build_runner entirely (or possibly just use it if present in deps). The main issue is the debugging support, there is a fair bit of work to get that all hooked up and tested etc.

> In general, I'm just hoping that we could get to a point where there's less differentiation in the DX for web-based projects than there is now.

Ya, I don't think anybody would disagree with that :). But we are unfortunately a pretty long ways from it.

**Comment by evanweible-wf** [comment]
That makes sense, thanks for the info! Just wanted to throw it out there in case there were other options for implementing this. Even if the process is different for web tests and VM tests, just being able to debug web tests from VSCode would be awesome, so I'm all for doing whatever works for now.
